LibraBFT/Abstract/BFT.agda

```diff
@@ -6,11 +6,479 @@
+            ∸-suc-≤ (suc x) (suc w) = ∸-suc-≤ x w
+
+
+
+
+   quorumInt>biz : ∀ (xs ys : List Member)
+                    → QSize ≤ length xs
+                    → QSize ≤ length ys
+                    → length xs + length ys ∸ authorsN ≤ length (intersect xs ys)
+                    → bizF + 1 ≤ length (intersect xs ys)
+   quorumInt>biz xs ys q≤x q≤y ≤int =
+     let p₁ = ≤-trans (∸-monoˡ-≤ authorsN (+-mono-≤ q≤x q≤y)) ≤int
+         p₂ = subst (_≤ length (intersect xs ys)) (simpExp₁ authorsN bizF) p₁
+         p₃ = ≤-trans (∸-monoˡ-≤ (2 * bizF) isBFT) p₂
+     in subst (_≤ length (intersect xs ys)) (simpExp₂ bizF) p₃
+       where  simpExp₁ : ∀ (x y : ℕ) → (x ∸ y) + (x ∸ y) ∸ x ≡ x ∸ (2 * y)
+              simpExp₁ x y rewrite sym (*-identityʳ (x ∸ y))
+                                 | sym (*-distribˡ-+ (x ∸ y) 1 1)
+                                 | *-comm (x ∸ y) 2
+                                 | *-distribˡ-∸ 2 x y
+                                 | ∸-+-assoc (2 * x) (2 * y) x
+                                 | +-comm (2 * y) x
+                                 | sym (∸-+-assoc (2 * x) x (2 * y))
+                                 | +-identityʳ x
+                                 | m+n∸n≡m x x = refl
+
+              simpExp₂ : ∀ (x : ℕ) → suc (3 * x) ∸ 2 * x ≡ x + 1
+              simpExp₂ x rewrite +-∸-assoc 1 (*-monoˡ-≤ x {2} {3} (s≤s (s≤s z≤n)))
+                               | sym (*-distribʳ-∸ x 3 2)
+                               | sym (+-suc x 0) = refl
+
+
+
+   span-hon : ∀ {xs dis hon : List Member} {x : Member}
+              → span Meta-dishonest? xs ≡ (dis , x ∷ hon)
+              → x ∈ xs ×  Meta-Honest-PK (getPubKey x)
+   span-hon {x₁ ∷ []} {dis} {hon} {x} eq
+     with Meta-dishonest? x₁
+   span-hon {x₁ ∷ []} {dis} {hon} {x} () | yes prf
+   span-hon {x₁ ∷ []} {[]} {[]} {x₁} refl | no ¬dis = here refl , ¬dis
+   span-hon {x₁ ∷ xs} {dis} {hon} {x} eq
+     with Meta-dishonest? x₁
+   span-hon {x₁ ∷ xs} refl | no ¬dis = (here refl) , ¬dis
+   ... | yes prf
+     with span Meta-dishonest? xs | inspect (span Meta-dishonest?) xs
+   span-hon {x₁ ∷ xs} refl | yes prf | fst , x₂ ∷ snd | [ eq₁ ]
+     = ×-map₁ there (span-hon {xs} eq₁)
+
+
+
+
+   span-dis : ∀ {xs dis : List Member}
+              → span Meta-dishonest? xs ≡ (dis , [])
+              → length xs ≡ length (List-filter Meta-dishonest? xs)
+   span-dis {[]} {dis} eq = refl
+   span-dis {x ∷ xs} {dis} eq
+     with Meta-dishonest? x | eq
+   ... | no ¬dis  | ()
+   ... | yes prf  | _
+     with span Meta-dishonest? xs | inspect (span Meta-dishonest?) xs
+   ... | fst , [] | [ eq₁ ] = cong suc (span-dis {xs} eq₁)
+
+
+
+   find-honest : ∀ {xs : List Member}
+                    → IsSorted _<Fin_ xs
+                    → bizF + 1 ≤ length xs
+                    → ∃[ α ] (α ∈ xs × Meta-Honest-PK (getPubKey α))
+   find-honest {xs} sxs biz<
+     with span Meta-dishonest? xs | inspect (span Meta-dishonest?) xs
```

Comment:
I am curious as to why you used `span` here.  I suspect this could be simpler, but am out of time to think more about how.  Easy to say, right :smile:.

Reply:
It seemed simpler to me. An alternative would be: first prove that the `length (List-filter Meta-dishonest? xs) ≤ length xs` and then prove that if it's `_<_` then `∃[ α ] (α ∈ xs × Meta-Honest-PK (getPubKey α))`. Otherwise, that is, if `length xs ≡ length (List-filter Meta-dishonest? xs)` we get a contradiction using the bft assumption (as we have now). I need to prove 2 auxiliary lemmas anyway. But at the moment using the span was more intuitive to me.

Reply:
That is how I would have approached it (but proving more general versions of those lemmas; in fact, the first one is in the library as `length-filter`).  But your new version of the proof of `span-hon` is much nicer and simpler (than both your initial version and my version I gave in feedback), so let's stick with this for now.

Reply:
But can I ask you one more thing: put a `TODO-1` comment suggesting that the proof can likely be simplified and briefly describe what you (and I) said above?  Once you do that, I think we can merge.  Thanks!

Reply:
Done! Thank you for the feedback!